### PR TITLE
Add cube map texture support.

### PR DESCRIPTION
**Pull request by rongguodong**
cubemap_shader uses the cube map texture in "sky box" mode.
cubemap_reflection_shader uses the cube map texture in reflection mode.

The new sample "gvr-cubemap" demonstrates both modes.

GearVRf-DCO-1.0-Signed-off-by: Guodong Rong
rongguodong@hotmail.com

### CODE REVIEW

**Comment by rongguodong** [comment]
Since loading a cube map texture asynchronously is not trivial, the new sample uses synchronous version. It is just to show how to use a cube map texture in both sky box mode and reflection mode. We can create asynchronous loading for cube map texture later, and modify the sample accordingly.

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 1]
```diff
@@ -489,6 +493,22 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
...
+
+        List<Bitmap> bitmapList = new ArrayList<Bitmap>(6);
+        for (GVRAndroidResource resource : resourceList) {
+            bitmapList.add(GVRAsynchronousResourceLoader.decodeStream(
+                    resource.getStream(), false));
+            resource.closeStream();
+        }
+
+        return resourceList.size() != 6 ? null : new GVRCubemapTexture(this,
+                bitmapList);
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: I think the size check should be at the beginning other wise it may be a performance waste when the texture is not valid.

File: GVRf/Framework/src/org/gearvrf/GVRMaterial.java [hunk 2]
```diff
@@ -97,7 +97,17 @@
...
+        public abstract static class Cubemap {
+            public static final GVRMaterialShaderId ID = new GVRStockMaterialShaderId(
+                    6);
+        }
+
+        public abstract static class CubemapReflection {
+            public static final GVRMaterialShaderId ID = new GVRStockMaterialShaderId(
+                    7);
+        }
+};
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/GVRMaterial.java]: These two classes should be inside GVRShaderType.

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 3]
```diff
@@ -341,6 +341,10 @@ public GVRMesh createQuad(float width, float height) {
                 width * 0.5f, height * -0.5f, 0.0f };
         mesh.setVertices(vertices);
 
+        float[] normals = { 0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 1.0f, 0.0f, 0.0f,
+                1.0f, 0.0f, 0.0f, 1.0f };
+        mesh.setNormals(normals);
+
         float[] texCoords = { 0.0f, 0.0f, 0.0f, 1.0f, 1.0f, 0.0f, 1.0f, 1.0f };
         mesh.setTexCoords(texCoords);
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: Just my personal opinion, since these normals and texCoords are constant values. I think we'd better to make them as final static member of this class.

**Comment by rongguodong** [comment]
I have modified code according to Chao and Jon's comments. Particularly, I added asynchronous loading support for cube map texture and use asynchronous loading in the new sample now.

I have also fixed bugs in old asynchronous loading code, and rebased with Tom's change.

Please review and test the code again. Thanks!

**Comment by rongguodong** [comment]
Note on the sample code:
There are two sets of bitmaps (beach.zip and screenshot3d.zip). The current code uses beach.zip for cube map texture since it is more beautiful. The commented out code uses screenshot3d.zip and uses the setFaceNames method.

File: GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        protected Bitmap[] loadResource() {
+            Bitmap[] bitmapArray = new Bitmap[6];
+            ZipInputStream zipInputStream = new ZipInputStream(
+                    resource.getStream());
+
+            try {
+                ZipEntry ze = null;
+                while ((ze = zipInputStream.getNextEntry()) != null) {
+                    String iamgeName = ze.getName();
+                    Integer imageIndex = GVRCubemapTexture.faceIndexMap
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: I think it should be imageName.

Comment (rongguodong) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: Fixed.

File: GVRf/Framework/jni/objects/textures/cubemap_texture.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+#include <android/bitmap.h>
+
+#include "objects/textures/texture.h"
+#include "util/gvr_log.h"
+
+namespace gvr {
+class CubemapTexture: public Texture {
+public:
+	explicit CubemapTexture(JNIEnv* env, jobjectArray bitmapArray) :
```
Comment (JonShemitz) [inline GVRf/Framework/jni/objects/textures/cubemap_texture.h]: Everywhere else, the `_jni.cpp` files contain all the JNI code, and make calls into the 'straight C++ code'. I think it would be best to follow that convention, here: do all the `env->` in `cubemap_texture_jni.cpp` and just pass in an array of six structs with the `AndroidBitmapInfo info` and the `void* pixels`.

File: GVRf/Framework/jni/shaders/material/cubemap_reflection_shader.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+// OpenGL Cube map texture uses coordinate system different to other OpenGL functions:
+// Positive x pointing right, positive y pointing up, positive z pointing inward.
+// It is a left-handed system, while other OpenGL functions use right-handed system.
+// The side faces are also oriented up-side down as illustrated below.
+//
+// Since the origin of Android bitmap is at top left, and the origin of OpenGL texture
+// is at bottom left, when we use Android bitmap to create OpenGL texture, it is already
+// up-side down. So we do not need to flip them again.
+//
+// We do need to flip the z-coordinate to be consistent with the left-handed system.
```
Comment (JonShemitz) [inline GVRf/Framework/jni/shaders/material/cubemap_reflection_shader.cpp]: Nice!

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 7]
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
     }
 
     /**
+     * Loads a cube map texture synchronously.
+     * 
+     * <p>
+     * Note that this method may take hundreds of milliseconds to return: unless
+     * the bitmap is quite tiny, you probably don't want to call this directly
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: `bitmap` s/b `cube map`

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 8]
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
...
+     * 
+     * @since 1.6.9
+     */
+    public GVRCubemapTexture loadCubemapTexture(
+            GVRAndroidResource[] resourceArray) {
+
+        assertGLThread();
+
+        if (resourceArray.length != 6) {
+            return null;
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: Should throw an `IllegalArgumentException`

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 9]
```diff
@@ -490,6 +495,44 @@ public GVRBitmapTexture loadTexture(GVRAndroidResource resource) {
...
+        }
+
+        Bitmap[] bitmapArray = new Bitmap[6];
+        for (int i = 0; i < 6; i++) {
+            bitmapArray[i] = GVRAsynchronousResourceLoader.decodeStream(
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: I probably ought to move `decodeStream()` to the utility package.

File: GVRf/Framework/src/org/gearvrf/GVRContext.java [hunk 10]
```diff
@@ -1093,6 +1136,29 @@ public void loadTexture(TextureCallback callback,
...
+     * {@link GVRShaders#setTexture(String, Future)}.
+     * 
+     * @param resource
+     *            A steam containing a zip file which contains six bitmaps. The
+     *            six bitmaps correspond to +x, -x, +y, -y, +z, and -z faces of
+     *            the cube map texture respectively. The default names of the
+     *            six images are "posx.png", "negx.png", "posy.png", "negx.png",
+     *            "posz.png", and "negz.png", which can be changed by calling
+     *            {@link GVRCubemapTexture#setFaceNames(String[])}.
+     * @returnA {@link Future} that you can pass to methods like
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRContext.java]: `@returnA` s/b `@return A`

File: GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import android.graphics.Bitmap;
+
+/** Cube map texture. */
+public class GVRCubemapTexture extends GVRTexture {
+    /**
+     * Constructs a cube map texture using a pre-existing {@link Bitmap}s.
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java]: `a` s/b `6`

File: GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+     * different to the default ones, this function must be called before load
+     * the zip file.
+     * 
+     * @param nameArray
+     *            An array containing six strings which are names of images
+     *            corresponding to +x, -x, +y, -y, +z, and -z faces of the cube
+     *            map texture respectively.
+     */
+    public static void setFaceNames(String[] nameArray) {
+        if (nameArray.length != 6) {
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java]: Should be using Chao's `Preconditions` (aka `Assert`) class.

File: GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+        for (int i = 0; i < 6; i++) {
+            faceIndexMap.put(nameArray[i], i);
+        }
+    }
+
+    /**
+     * Internal map data structure. Public only for visible to
+     * {@code AsyncCubemapTexture}.
+     */
+    public static Map<String, Integer> faceIndexMap = new HashMap<String, Integer>(
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/GVRCubemapTexture.java]: * Should probably be `final` 
* Is there a way to make it `private` (or package-visible) and explicitly pass it to `AsyncCubemapTexture`?

File: GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+                resource, priority);
+    }
+
+    /*
+     * Static constants
+     */
+
+    private static final String TAG = Log.tag(AsyncCubemapTexture.class);
+
+    private static final Class<? extends GVRHybridObject> TEXTURE_CLASS = GVRTexture.class;
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: I'm pretty sure this breaks asynch textures - you need to register `AsyncCubemapTexture` here.

File: GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+                    resource.getStream());
+
+            try {
+                ZipEntry zipEntry = null;
+                while ((zipEntry = zipInputStream.getNextEntry()) != null) {
+                    String imageName = zipEntry.getName();
+                    Integer imageIndex = GVRCubemapTexture.faceIndexMap
+                            .get(imageName);
+                    if (imageIndex == null) {
+                        throw new IllegalArgumentException("Name of image ("
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: The `Exceptions` class would let you use `String.format()` syntax.

File: GVRf/Framework/src/org/gearvrf/asynchronous/GVRAsynchronousResourceLoader.java [hunk 16]
```diff
@@ -299,6 +301,44 @@ public void run() {
...
+     *            A steam containing a zip file which contains six bitmaps. The
+     *            six bitmaps correspond to +x, -x, +y, -y, +z, and -z faces of
+     *            the cube map texture respectively. The default names of the
+     *            six images are "posx.png", "negx.png", "posy.png", "negx.png",
+     *            "posz.png", and "negz.png", which can be changed by calling
+     *            {@link GVRCubemapTexture#setFaceNames(String[])}.
+     * @param priority
+     *            This request's priority. Please see the notes on asynchronous
+     *            priorities in the <a href="package-summary.html#async">package
+     *            description</a>. Also, please note priorities only apply to
```
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/asynchronous/GVRAsynchronousResourceLoader.java]: Zap `Also, please note ... request scheduler.`

**Comment by JonShemitz** [comment]
Looks really good. Aside from
1. The JNI code outside of a `_jnp.cpp` class and
2. The probable mis-registration in `AsyncCubemapTexture.java`
all my comments are really pretty minor.

File: GVRf/Framework/jni/objects/textures/cubemap_texture.h (hunk 5 above)
Comment (rongguodong) [inline GVRf/Framework/jni/objects/textures/cubemap_texture.h]: I moved env->GetArrayLength(bitmapArray) to cubemap_texture_jni.h.

For env->GetObjectArrayElement and AndroidBitmap_* functions, I think it is more efficient to be in cubemap_texture.h. If we move all of them to cubemap_texture_jni.h, we need to create six objects there and pass them to cubemap_texture.h, and delete the six objects then. This will introduce necessary memory allocation and delocation. I think it is the reason why base_texture.h has all AndroidBitmap_* functions in it, not in base_texture_jni.h.

Comment (JonShemitz) [inline GVRf/Framework/jni/objects/textures/cubemap_texture.h]: > If we move all of them to cubemap_texture_jni.h, we need to create six objects there and pass them to cubemap_texture.h, and delete the six objects then. This will introduce necessary memory allocation and delocation. 

Allocating and deallocating a local array of `struct`s is very cheap ,,, but the existence of this grody code in `base_texture.cpp` sort of undercuts my Follow The Format argument, here.

**Comment by rongguodong** [comment]
I pushed a new commit addressing Jon's comments. For condition checking parts, since we are modifying the Assert class, I can modify them after #44 are merged.

File: GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+                CancelableCallback<GVRHybridObject> callback, int priority) {
+            super(gvrContext, sConverter, request, callback);
+        }
+
+        @Override
+        protected Bitmap[] loadResource() {
+            Bitmap[] bitmapArray = new Bitmap[6];
+            ZipInputStream zipInputStream = new ZipInputStream(
+                    resource.getStream());
+
```
Comment (chenchao1407) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: Just a minor concern here.

Forcing developers to use zip to construct a cubetexture may lose some flexibility for some cases e.x. : if one want to use several combinations of same bunch of textures.

**Comment by chenchao1407** [comment]
Just reviewed the codes. Looks fine to me aside from some minor concerns.

Tested. Runs well. Good to go merged if there is no any other objection.

Do you have any other comment? @JonShemitz

File: GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java (hunk 17 above)
Comment (JonShemitz) [inline GVRf/Framework/src/org/gearvrf/asynchronous/AsyncCubemapTexture.java]: > several combinations of same bunch of textures.

Perhaps you and Guodong discussed this in-person? From what I know of cube maps, this doesn't seem like a very likely scenario: while I'm sure you could come up with examples (manufactured objects with pre-fab, snap-in panels?) I gather that most cube maps are renders (or captures) of complex objects; most of the time, each face will be a unique texture. I may (of course!) be wrong.
